docs/networking.adoc

```diff
@@ -0,0 +1,766 @@
+
+IPv6 addresses can be used in most places that an IPv4 address can be.  That includes command-line tools like `ping(8)` and `dig(1)`.  If you're using it where something expects and IP and port, as in an HTTP URL, you generally put the IP address in square brackets: `http://[fdb0:a840:2504:3d5::1]:8080`.
+
+=== IPv6 address space
+
+The IPv6 address space is huge.  Many blocks are assigned specific purposes.  Some that are notable for us:
+
+* `fe80::/10`: https://en.wikipedia.org/wiki/Link-local_address[Link-local addresses].  Every IPv6 interface has one of these and it's assigned automatically without DHCP or anything.  However, these addresses are only valid on the network segment that the interface is attached to.
+* `fc00::/7`: https://en.wikipedia.org/wiki/Unique_local_address[Unique local address (ULA)], "somewhat analogous to IPv4 private network addressing" (e.g., `10.0.0.0/8`).
+** `fd00::/8`: the part of the ULA that's _not_ centrally assigned.  The control plane primarily uses addresses in this range.
+
+Because link-local addresses are only valid for the network that the link is attached to, whenever you use a link-local address, you usually have to specify which link's network you mean.  You might see `fe80::aa40:25ff:fe04:3d5%cxgbe0`, where `fe80::aa40:25ff:fe04:3d5` is the link-local address and `cxgbe0` is the **scope_id** (denoting "the network attached to cxgbe0").  We'll look at an example below.
+
+== Inspecting networking configuration on illumos
+
+illumos supports lightweight OS-based virtualization in the form of **zones**.  Each system has a **global zone** (GZ) that contains all other **non-global zones.**  Each zone looks like its own separate illumos system with its own root filesystem and (importantly for us) its own netstack.  It has its own set of datalinks, IP interfaces, addresses, etc.
+
+Let's start at the link layer (L2) in the GZ of a sled.  https://illumos.org/man/8/dladm[dladm(8)] shows all the L2 links in the system:
+
+[source,console]
+----
+BRM44220011 # dladm
+LINK        CLASS     MTU    STATE    BRIDGE     OVER
+tfpkt0      phys      1500   down     --         --
+cxgbe0      phys      9000   up       --         --
+cxgbe1      phys      9000   up       --         --
+bootstrap_stub0 etherstub 9000 up     --         --
+bootstrap0  vnic      1500   up       --         bootstrap_stub0
+underlay_stub0 etherstub 9000 up      --         --
+underlay0   vnic      9000   up       --         underlay_stub0
+----
+
+We can see:
+
+* Three `phys` links that correspond to physical devices.  We're interested in `cxgbe0` and `cxgbe1`, which correspond to the two ports provided by the Chelsio NIC on Gimlets.
+* Two `etherstub` devices, `bootstrap_stub0` and `underlay_stub0`.  Etherstubs function like a synthetic Ethernet network.  We can create VNICs over an etherstub and those VNICs will act as though connected through a (virtual) switch.
```

Comment:
Makes sense.  I'm not sure a better way to explain this, though.  Feel free to take a stab!

Reply:
Also, I only realized this later on, but `etherstub`s are actually just `VNIC`s without an underlying link

Reply:
Maybe adding something to the VNIC section like:

When you create a vnic on top of a device, it implicitly is creating an internal (virtual) L2 switch. Put another way creating a VNIC, whether on top of a physical device, etherstub, or other types of data links, can be imagined as plugging the VNICs into an unmanaged switch with the other device as the uplink. When traffic is not destined for something within the switch it will always go out the uplink device.

Reply:
After some discussion offline with @rmustacc I understand this better (along with @luqmana's comment!).  I've expanded this section a bit and added some diagrams based on an ASCII diagram from Robert.  It might be a little distracting given where it is in the flow, but I think it's pretty valuable information.